libs/prebuilt/langgraph/prebuilt/tool_node.py

```diff
@@ -328,24 +376,64 @@ def __init__(
         """Initialize the ToolNode with the provided tools and configuration.
 
         Args:
-            tools: Sequence of tools to make available for execution.
+            tools: Either a sequence of tools to make available for execution,
+                  or a callable that returns tools dynamically based on input,
+                  config, and store.
             name: Node name for graph identification.
             tags: Optional metadata tags.
             handle_tool_errors: Error handling configuration.
             messages_key: State key containing messages.
         """
         super().__init__(self._func, self._afunc, name=name, tags=tags, trace=False)
-        self.tools_by_name: dict[str, BaseTool] = {}
-        self.tool_to_state_args: dict[str, dict[str, Optional[str]]] = {}
-        self.tool_to_store_arg: dict[str, Optional[str]] = {}
         self.handle_tool_errors = handle_tool_errors
         self.messages_key = messages_key
-        for tool_ in tools:
-            if not isinstance(tool_, BaseTool):
-                tool_ = create_tool(tool_)
-            self.tools_by_name[tool_.name] = tool_
-            self.tool_to_state_args[tool_.name] = _get_state_args(tool_)
-            self.tool_to_store_arg[tool_.name] = _get_store_arg(tool_)
+
+        if callable(tools):
+            # Dynamic tool provider
+            self._tool_provider_fn = tools
+            self._static_resolver: Optional[ToolResolver] = None
+            # Likely migrate to property and raise a RunTimeError
+            self.tools_by_name: dict[str, BaseTool] = {}
```

Comment:
TODO: tools_by_name needs to be cleaned up, likely driven using a property. We can't delete since it's part of the public API